### ISSUE TITLE
Bicep parameters with default values are always passed as null

### ISSUE THREAD

**Issue by TWolversonReply** [Azure/azure-dev#4206]
- [ ] Make sure you've installed the latest version using [instructions in the wiki](../wiki/install)

**Output from `azd version`**
azd version 1.9.5 (commit cd2b7af9995d358aab33c782614f801ac1997dde)

**Describe the bug**
Bicep parameters with default values are always passed as null, regardless of config.json

**To Reproduce**
Steps to reproduce the behavior...

**Expected behavior**
Define a parameter with a default value in main.bicep
`param clientTags string = '{}'`

azd provision doesn't prompt for it, which is expected, but the deployment behaves as though the param is null, regardless of what the default value is actually set to. 

Crucially, supplying a value in config.json for the current environment makes no difference, the parameter is still null.

**Environment**
Information on your environment:
    * Language name and version
    * IDE and version : [e.g. Visual Studio 16.3]

**Additional context**

**Comment by rajeshkamal5050** [comment]
@jongio @vhvb1989 don't think azd passes/translates such default empty string params as null for deployment? any tips/tricks to handle it in bicep itself?

**Comment by ChrisL-GU** [comment]
I'm running into this issue, supplying a value in config.json still results in the optional parameter always being set to null.

In some environments I want the value to be used in the Bicep and in other environments when the value is not specified then the operation should not be performed in the Bicep. Config.json seemed like a nice way to supply or not supply the value. Since it seems very few people are running into this I must be thinking about the scenario incorrectly. 🤔

**Comment by vhvb1989** [comment]
Hi @ChrisL-GU , could you provide more details about your use-case?

**Comment by ChrisL-GU** [comment]
Thanks for following up @vhvb1989 

I'm deploying a Web App which connects to a Cosmos database. When the AZD environment is Dev or Test then I want my personal Entra ID user to have access to the database so I can run the application locally. But for Production I do not want my personal Entra ID user to have access. I'm using this Bicep to grant my personal Entra ID user access:

```
resource cosmosDbSqlRoleAssignmentLocalDev 'Microsoft.DocumentDB/databaseAccounts/sqlRoleAssignments@2023-11-15' = if (<some conditional here>) {
  name: guid(cosmosDbAccount.id, localDevUserObjectId, 'sqlDataReader')
  parent: cosmosDbAccount
  properties: {
    roleDefinitionId: cosmosDbSqlDataReaderRoleDefinitionId
    principalId: localDevUserObjectId
    scope: '${cosmosDbAccount.id}/dbs/${cosmosDbDatabaseName}'
  }
}
```

I was hoping in the Dev and Test AZD environment I could include the parameter `localDevUserObjectId` in `config.json` with the value of my Id, but not include this parameter in the Production AZD environment. Then the parameter in the Bicep could be optional:

```
@description('Object ID of the Entra user for local development')
@minLength(36)
@maxLength(36)
param localDevUserObjectId string?
```

But that doesn't seem to work, the value will not be read from `config.json` if the param is optional.

**Comment by vhvb1989** [comment]
it looks like you are using `localDevUserObjectId` to distinguish between PROD (using its default value) or non-PROD (setting a value for it).

Instead of making the parameter nullable with `param localDevUserObjectId string?`, you might want to try setting an empty default value like `param localDevUserObjectId string = ''` 

Then you can build your condition about `if localDevUserObjectId is empty -> skip the EntraId bicep module` 

Then, next to the `main.bicep` file, you should have a `main.parameters.json` . There you would define an ENV VAR which you would use to override the parameter, like:

<img width="598" height="205" alt="Image" src="https://github.com/user-attachments/assets/21f3d98e-f42f-43b5-9b64-ba88072a8d5d" />

This would help you to set that ENV VAR in your DEV and TEST envs.

NOTE:  Avoid linking the ENV VAR to `${AZURE_PRINCIPAL_ID}`, as that would be always set by AZD regardless of the ENV.

Some folks would use an input parameter like `param isProduction bool = false` as a simpler switch.  That parameter would make you use whatever principal id you get from AZD depending on its value.

**Comment by ChrisL-GU** [comment]
Thank you for the tips. I think the key piece I wasn't aware of was the usage of `${AZURE_PRINCIPAL_ID}`. Since I don't have to provide this value myself I can use its provided value and then add a switch param, similar to what you shared.

However, I cannot set the switch param to a default value. since default values make parameters optional, meaning the value I use in the `config.json` will not be read. But that's okay for my scenario, I can make the parameter required and set it in every `config.json`.

I believe this is a better solution than what I originally had put together. Although, it's still disappointing, and surprising, to me that `config.json` values will not be read if parameters are optional.

**Comment by vhvb1989** [comment]
When you say `config.json` , are you referring to `.azure/env-name/config.json` ?  or are you referring to `main.parameters.json` ?
The former is mainly implementation details for AZD. It is used to persist your manual selections from the prompt.  
The expectation is users to link ENV VARS to input parameters, without knowing anything about the config file form the environment.

**Comment by ChrisL-GU** [comment]
Yes, `azure/env-name/config.json`. The `config.json` seemed like a more resilient location to preset parameters, and isolate those values. But I see, I'm abusing the intent of the file.

**Comment by TWolversonReply** [comment]
I've misstated the issue, let me try and approach this again;

Given a `main.parameters.json`:

```
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
     "notificationTopicName": {
      "value": "${NOTIFICATION_TOPIC_NAME}"
    }
}
```

and a `main.bicep`:

```
@description('Name of the topic, where workflow publish the notifications ')
param notificationTopicName string = 'notification'
```

if the environment variable NOTIFICATION_TOPIC_NAME isn't set, the mapping in parameters.json is interpreted as **empty string** (not null), which is a valid value for a string parameter, and the default doesn't take effect.

I consider it fairly obvious that the intended behaviour would be that if the environment variable representing this parameter isn't set, the default is taken instead. This isn't what happens. How are you supposed to work around this?

**Comment by vhvb1989** [comment]
@TWolversonReply - that's how it works.

Here's an example. Using:
 - paramDefault:  Name of the bicep input param
 - PARAM_RESULT:  output param to test the resolved value
 - AZURE_PARAM_DEFAULT:  env var mapped to paramDefault

If `AZURE_PARAM_DEFAULT` is not defined in your env:

<img width="1677" height="540" alt="Image" src="https://github.com/user-attachments/assets/5f7a355a-f45e-4f76-87e5-fe0cadf239d0" />

The resolved value is the bicep input parameter's default value.

Then if I define `AZURE_PARAM_DEFAULT` in my env (export AZURE_PARAM_DEFAULT=""), azd will take whatever value is in the env var for the parameter. Either an empty string or some string.

<img width="1812" height="542" alt="Image" src="https://github.com/user-attachments/assets/68e7a3cb-851b-406b-b087-807798fd2246" />

In the past (old azd versions), azd used to consider env vars with empty strings to be the same as non-defined env vars, blocking the use case of setting empty strings values with env vars. This was a bug and was fixed it.

If you want to override the default value from the main.bicep without actually changing its value, you can use main.parameters.json like this:

<img width="1825" height="534" alt="Image" src="https://github.com/user-attachments/assets/6e0e10f8-e407-4380-89a9-2f389c835a0c" />

Here you can see:
```json
"paramDefault":{
        "value": "${AZURE_PARAM_DEFAULT='override'}"
      }
```

This means that you might be using a main.bicep for a base case with fundamental default but then you might have multiple parameters file which changes the default value for the parameter.  If the ENV VAR is not defined, the parameters file decides what's the default value to use.  Then, if you switch back to another parameters file for deploying with no default-value override, it would fall back to what main.bicep defines as default value for the parameter.

**Comment by TWolversonReply** [comment]
@vhvb1989 I retested with a minimal repro yesterday and agree that this is how I've observed it to work now, but the behaviour must have changed since I raised the issue.

As an aside, I have never seen this syntax before:
```
"paramDefault":{
        "value": "${AZURE_PARAM_DEFAULT='override'}"
      }
```
and can't find any reference to it in the documentation.

**Comment by vhvb1989** [comment]
That makes sense, @TWolversonReply .  I can see the issue was created for azd version 1.9.5
This is the change: https://github.com/Azure/azure-dev/pull/5061
It was released for version 1.15.0

> Making AZD to ignore parameters mapped to ENV VARS when the ENV VAR is not defined. In the past AZD was taking empty string as a value for these mappings.

The syntax you mentioned has always been supported but you are correct, it is not documented 😔   
I have created https://github.com/Azure/azure-dev/issues/5534 to fix that.  Thank you for mentioning it 😊
